File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/timed/DeleteDisabledUsersServiceTest.java
package uk.gov.justice.digital.hmpps.oauth2server.timed;

import com.microsoft.applicationinsights.TelemetryClient;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserRetries;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken.TokenType;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRetriesRepository;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class DeleteDisabledUsersServiceTest {
    @Mock
    private UserRepository userRepository;
    @Mock
    private UserRetriesRepository userRetriesRepository;
    @Mock
    private TelemetryClient telemetryClient;
    @Captor
    private ArgumentCaptor<Map<String, String>> mapCaptor;

    private DeleteDisabledUsersService service;

    @Before
    public void setUp() {
        service = new DeleteDisabledUsersService(userRepository, userRetriesRepository, telemetryClient);
    }

    @Test
    public void findAndDeleteDisabledUsers_Processed() {
        final var users = List.of(User.of("user"), User.of("joe"));
        when(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
                .thenReturn(users);
        assertThat(service.processInBatches()).isEqualTo(2);
    }

    @Test
    public void findAndDeleteDisabledUsers_Deleted() {
        final var user = User.of("user");
        final var joe = User.of("joe");
        final var users = List.of(user, joe);

        when(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
                .thenReturn(users);
        service.processInBatches();
        verify(userRepository).delete(user);
        verify(userRepository).delete(joe);
    }

    @Test
    public void findAndDeleteDisabledUsers_DeleteAll() {
        final var user = User.of("user");
        final var token = user.createToken(TokenType.RESET);
        final var retry = new UserRetries("user", 3);
        when(userRetriesRepository.findById(anyString())).thenReturn(Optional.of(retry));

        when(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
                .thenReturn(List.of(user));
        service.processInBatches();
        verify(userRetriesRepository).delete(retry);
    }

    @Test
    public void findAndDeleteDisabledUsers_Telemetry() {
        final var users = List.of(User.of("user"), User.of("joe"));
        when(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
                .thenReturn(users);
        service.processInBatches();

        verify(telemetryClient, times(2)).trackEvent(eq("DeleteDisabledUsersProcessed"), mapCaptor.capture(), isNull());
        assertThat(mapCaptor.getAllValues().stream().map(m -> m.get("username")).collect(Collectors.toList())).containsExactly("user", "joe");
    }
}

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/timed/DeleteDisabledUsersServiceTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.timed

import com.microsoft.applicationinsights.TelemetryClient
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Captor
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserRetries
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRetriesRepository
import java.util.*

@ExtendWith(MockitoExtension::class)
class DeleteDisabledUsersServiceTest {
  private val userRepository: UserRepository = mock()
  private val userRetriesRepository: UserRetriesRepository = mock()
  private val telemetryClient: TelemetryClient = mock()

  @Captor
  private lateinit var mapCaptor: ArgumentCaptor<Map<String, String>>
  private val service: DeleteDisabledUsersService = DeleteDisabledUsersService(userRepository, userRetriesRepository, telemetryClient)

  @Test
  fun findAndDeleteDisabledUsers_Processed() {
    val users = listOf(User.of("user"), User.of("joe"))
    whenever(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
        .thenReturn(users)
    assertThat(service.processInBatches()).isEqualTo(2)
  }

  @Test
  fun findAndDeleteDisabledUsers_Deleted() {
    val user = User.builder().username("user").id(UUID.randomUUID()).build()
    val joe = User.builder().username("joe").id(UUID.randomUUID()).build()
    val users = listOf(user, joe)
    whenever(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
        .thenReturn(users)
    service.processInBatches()
    verify(userRepository).delete(user)
    verify(userRepository).delete(joe)
  }

  @Test
  fun findAndDeleteDisabledUsers_DeleteAll() {
    val user = User.of("user")
    user.createToken(UserToken.TokenType.RESET)
    val retry = UserRetries("user", 3)
    whenever(userRetriesRepository.findById(anyString())).thenReturn(Optional.of(retry))
    whenever(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
        .thenReturn(listOf(user))
    service.processInBatches()
    verify(userRetriesRepository).delete(retry)
  }

  @Test
  fun findAndDeleteDisabledUsers_Telemetry() {
    val users = listOf(User.of("user"), User.of("joe"))
    whenever(userRepository.findTop10ByLastLoggedInBeforeAndEnabledIsFalseOrderByLastLoggedIn(any()))
        .thenReturn(users)
    service.processInBatches()
    verify(telemetryClient, times(2)).trackEvent(ArgumentMatchers.eq("DeleteDisabledUsersProcessed"), mapCaptor.capture(), ArgumentMatchers.isNull())
    assertThat(mapCaptor.allValues.map { it["username"] }).containsExactly("user", "joe")
  }
}
